src/FirmwarePlugin/APM/APMParameterMetaData.cc

```diff
@@ -78,29 +79,396 @@ QVariant APMParameterMetaData::_stringToTypedVariant(const QString& string, Fact
+        qCWarning(APMParameterMetaDataLog) << "Badly formed XML, reading failed: " << xml.errorString();
+        return;
+    }
+
+    QString             errorString;
+    bool                badMetaData = true;
+    QStack<int>         xmlState;
+    APMFactMetaDataRaw* rawMetaData = NULL;
+
+    xmlState.push(XmlStateNone);
+
+    QMap<QString,QStringList> groupMembers; //used to remove groups with single item
+
+    while (!xml.atEnd()) {
+        if (xml.isStartElement()) {
+            QString elementName = xml.name().toString();
+
+            if (elementName.isEmpty()) {
+                // skip empty elements
+            } else if (elementName == "paramfile") {
+                if (xmlState.top() != XmlStateNone) {
+                    qCWarning(APMParameterMetaDataLog) << "Badly formed XML, paramfile matched";
+                }
+                xmlState.push(XmlstateParamFileFound);
+                // we don't really do anything with this element
+            } else if (elementName == "vehicles") {
+                if (xmlState.top() != XmlstateParamFileFound) {
+                    qCWarning(APMParameterMetaDataLog) << "Badly formed XML, vehicles matched";
+                    return;
+                }
+                xmlState.push(XmlStateFoundVehicles);
+            } else if (elementName == "libraries") {
+                if (xmlState.top() != XmlstateParamFileFound) {
+                    qCWarning(APMParameterMetaDataLog) << "Badly formed XML, libraries matched";
+                    return;
+                }
+                currentCategory = "libraries";
+                xmlState.push(XmlStateFoundLibraries);
+            } else if (elementName == "parameters") {
+                if (xmlState.top() != XmlStateFoundVehicles && xmlState.top() != XmlStateFoundLibraries) {
+                    qCWarning(APMParameterMetaDataLog) << "Badly formed XML, parameters matched";
+                    return;
+                }
+
+                if (xml.attributes().hasAttribute("name")) {
+                    // we will handle metadata only for specific MAV_TYPEs and libraries
+                    const QString nameValue = xml.attributes().value("name").toString();
+                    if (nameValue.contains(parameterCategories)) {
+                        xmlState.push(XmlStateFoundParameters);
+                        currentCategory = nameValue;
+                    } else if(xmlState.top() == XmlStateFoundLibraries) {
+                        // we handle all libraries section under the same category libraries
+                        // so not setting currentCategory
+                        xmlState.push(XmlStateFoundParameters);
+                    } else {
+                        qCDebug(APMParameterMetaDataLog) << "not interested in this block of parameters skip";
+                        if (skipXMLBlock(xml, "parameters")) {
+                            qDebug() << "something wrong with the xml, skip of the xml failed";
```

Comment:
assert was there just to give stall me unless problems are fixed. I will
definitely remove it once I send the actual PR.

On Sat, Dec 12, 2015 at 12:15 AM, pritam.ghanghas@gmail.com <
pritam.ghanghas@gmail.com> wrote:

> thats ok. I will add that but are you otherwise happy with the approach
> that you have taken?
>
> On Sat, Dec 12, 2015 at 12:08 AM, Don Gagne <notifications@github.com>
> wrote:
>
>> In src/FirmwarePlugin/APM/APMParameterMetaData.cc
>> <https://github.com/mavlink/qgroundcontrol/pull/2237#discussion_r47388107>
>> :
>>
>> > +                }
>> > +
>> > +                if (xml.attributes().hasAttribute("name")) {
>> > +                    // we will handle metadata only for specific MAV_TYPEs and libraries
>> > +                    const QString nameValue = xml.attributes().value("name").toString();
>> > +                    if (nameValue.contains(parameterCategories)) {
>> > +                        xmlState.push(XmlStateFoundParameters);
>> > +                        currentCategory = nameValue;
>> > +                    } else if(xmlState.top() == XmlStateFoundLibraries) {
>> > +                        // we handle all libraries section under the same category libraries
>> > +                        // so not setting currentCategory
>> > +                        xmlState.push(XmlStateFoundParameters);
>> > +                    } else {
>> > +                        qCDebug(APMParameterMetaDataLog) << "not interested in this block of parameters skip";
>> > +                        if (skipXMLBlock(xml, "parameters")) {
>> > +                            qDebug() << "something wrong with the xml, skip of the xml failed";
>>
>> Mission qCDebug
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/mavlink/qgroundcontrol/pull/2237/files#r47388107>.
>>
>
>
>
> --
> Regards,
> pritam
>



-- 
Regards,
pritam